core/src/serve_repair.rs

```diff
@@ -516,6 +524,18 @@ impl ServeRepair {
             inc_new_counter_debug!("serve_repair-handle-repair--eq", stats.self_repair);
         }
 
+        let staked_nodes = self.bank_forks.read().unwrap().root_bank().staked_nodes();
+        let (staked, unstaked): (Vec<_>, Vec<_>) = stats
+            .outgoing_bytes_requested
+            .iter()
+            .partition(|(id, _)| staked_nodes.contains_key(id));
+        let staked_response_bytes: u64 = staked.iter().map(|(_, x)| *x).sum();
+        let unstaked_response_bytes: u64 = unstaked.iter().map(|(_, x)| *x).sum();
+        let max_staked_peer_response_bytes: u64 =
+            staked.iter().map(|(_, x)| **x).max().unwrap_or_default();
+        let max_unstaked_peer_response_bytes: u64 =
+            unstaked.iter().map(|(_, x)| **x).max().unwrap_or_default();
```

Comment:
Yes, my plan is to implement limits, but I wanted to use the metrics to determine reasonable limits which wouldn't significantly impact the cluster. Once the limits are added these metrics can be removed.

Reply:
I understand that;
But my question was that if we want to prioritize requests by stake, do we still need these metrics first?
It seems to me that priority by stake will already enforce desired allocation.

Reply:
I don't think we currently know the breakdown of repair traffic from staked/unstaked nodes. We could prioritize staked over unstaked in a given batch using a bandwith limit derived from the batching time, but I assume this would impact traffic bursts which would currently be allowed.

Reply:
well, the issue with unstaked nodes is that you can generate an unlimited number of requests with random pubkeys so we can't allow that unstaked nodes' requests ever front run staked nodes' requests no matter how many unstaked nodes' requests we receive.

Regarding gathering metrics and some breakdowns, I guess an alternative would be to add the metrics to the node which _receives repair response_; i.e. submit metrics for the size of repair response packets that the node receives.

Then we can query the metrics from the whole cluster and we also know each node's stake. So we can aggregate across the cluster  to get some ballpark breakdown. The advantage is that there is no runtime overhead to lookup the stake and do the bookkeeping for metrics report.

The problem is that not all nodes submit metrics, but probably that wouldn't be too bad for our purposes here;

Reply:
I have a draft for changes to enforce bandwidth limits: #27771.

I still think these metrics would be temporarily useful to see outliers and inform the limits we want to enforce. I see that the existing bandwidth cap is being hit on some nodes on mainnet/testnet. I don't think performing this calculation once every 3s is particularly expensive compared to looking up multiple shreds from blockstore for a given request.

Reply:
Lets then first finalize the approach in #27771 and then see what kind of additional metrics, if any, are needed .